File: app/src/main/java/chat/rocket/android/layouthelper/chatroom/MessageFormManager.java
package chat.rocket.android.layouthelper.chatroom;

import chat.rocket.android.widget.message.MessageFormLayout;

/**
 * Handles MessageForm.
 */
public class MessageFormManager {
  private final MessageFormLayout messageFormLayout;
  private SendMessageCallback sendMessageCallback;

  public MessageFormManager(MessageFormLayout messageFormLayout, MessageFormLayout.ExtraActionSelectionClickListener callback) {
    this.messageFormLayout = messageFormLayout;
    init(callback);
  }

  private void init(MessageFormLayout.ExtraActionSelectionClickListener listener) {
    messageFormLayout.setExtraActionSelectionClickListener(listener);
    messageFormLayout.setSubmitTextListener(this::sendMessage);
  }

  public void setSendMessageCallback(SendMessageCallback sendMessageCallback) {
    this.sendMessageCallback = sendMessageCallback;
  }

  public void onMessageSend() {
    clearComposingText();
  }

  public void setEditMessage(String message) {
    clearComposingText();
    messageFormLayout.setText(message);
  }

  public void clearComposingText() {
    messageFormLayout.setText("");
  }

  private void sendMessage(String message) {
    if (sendMessageCallback == null) {
      return;
    }
    sendMessageCallback.onSubmitText(message);
  }

  public interface SendMessageCallback {
    void onSubmitText(String messageText);
  }
}

File: app/src/main/java/chat/rocket/android/layouthelper/chatroom/MessageFormManager.kt
package chat.rocket.android.layouthelper.chatroom

import android.text.TextUtils
import android.view.View
import chat.rocket.android.widget.AbsoluteUrl
import chat.rocket.android.widget.message.MessageFormLayout
import chat.rocket.core.models.Message

class MessageFormManager(private val messageFormLayout: MessageFormLayout, val callback: MessageFormLayout.ExtraActionSelectionClickListener) {
    private var sendMessageCallback: SendMessageCallback? = null
    private var replyMarkDown: String = ""

    init {
        messageFormLayout.setExtraActionSelectionClickListener(callback)
        messageFormLayout.setSubmitTextListener(this::sendMessage)
    }

    fun setSendMessageCallback(sendMessageCallback: SendMessageCallback) {
        this.sendMessageCallback = sendMessageCallback
    }

    fun onMessageSend() {
        clearComposingText()
    }

    fun setEditMessage(message: String) {
        clearComposingText()
        messageFormLayout.setText(message)
    }

    fun clearComposingText() {
        messageFormLayout.setText("")
    }

    fun enableComposingText(enable: Boolean) {
        messageFormLayout.isEnabled = enable
    }

    fun setReply(absoluteUrl: AbsoluteUrl, replyMarkDown: String, message: Message) {
        this.replyMarkDown = replyMarkDown
        messageFormLayout.setReplyContent(absoluteUrl, message)
        messageFormLayout.setReplyCancelListener({
            this.replyMarkDown = ""
            messageFormLayout.clearReplyContent()
        })
    }

    private fun sendMessage(message: String) {
        val finalMessage = if (replyMarkDown.isNotEmpty()) "$replyMarkDown $message" else message
        replyMarkDown = ""
        sendMessageCallback?.onSubmitText(finalMessage)
    }

    interface SendMessageCallback {
        fun onSubmitText(messageText: String)
    }
}